src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs

```diff
@@ -415,6 +422,152 @@ private void HandleCodeActionRequest(string id, string paramsJson)
+            var encodedTokens = SemanticTokensEncoder.EncodeTokens(tokens, expression, eol);
+
+            _sendToClient(JsonRpcHelper.CreateSuccessResult(id, new SemanticTokensResponse() { Data = encodedTokens }));
+        }
+
+        /// <summary>
+        /// Handles requests to compute semantic requests for the a specific part of document or expression.
+        /// </summary>
+        /// <param name="id">Request Id.</param>
+        /// <param name="paramsJson">Request Params Stringified Body.</param>
+        private void HandleRangeDocumentSemanticTokens(string id, string paramsJson)
+        {
+            if (!TryParseAndValidateSemanticTokenParams(id, paramsJson, out SemanticTokensRangeParams semanticTokensParams))
+            {
+                return;
+            }
+
+            if (semanticTokensParams.Range == null)
+            {
+                // No tokens for invalid range
+                SendEmptySemanticTokensResponse(id);
+                return;
+            }
+
+            var uri = new Uri(semanticTokensParams.TextDocument.Uri);
+            var queryParams = HttpUtility.ParseQueryString(uri.Query);
+            var expression = queryParams.Get("expression");
+
+            if (string.IsNullOrEmpty(expression))
+            {
+                // Empty tokens for the empty expression
+                SendEmptySemanticTokensResponse(id);
+                return;
+            }
+
+            // Monaco-Editor sometimes uses \r\n for the newline character. \n is not always the eol character so allowing clients to pass eol character
+            var eol = queryParams.Get("eol");
+            eol = !string.IsNullOrEmpty(eol) ? eol : EOL.ToString();
+
+            var (startIndex, endIndex) = PositionRangeHelper.ConvertRangeToPositions(semanticTokensParams.Range, expression, eol);
+            if (startIndex < 0 || endIndex < 0)
+            {
+                SendEmptySemanticTokensResponse(id);
+                return;
+            }
+
+            var tokenTypesToSkip = ParseTokenTypesToSkipParam(queryParams.Get("tokenTypesToSkip"));
+            var scope = _scopeFactory.GetOrCreateInstance(semanticTokensParams.TextDocument.Uri);
+            var result = scope.Check(expression);
+
+            // Skip over the token types that clients don't want in the response
+            var tokens = result.GetTokens().Where(tok => !tokenTypesToSkip.Contains(tok.TokenType));
+
+            // Only consider overlapping tokens. end index is exlcusive
+            var overlappingTokens = tokens.Where(token => !(token.EndIndex <= startIndex || token.StartIndex >= endIndex));
+            var encodedTokens = SemanticTokensEncoder.EncodeTokens(overlappingTokens, expression, eol);
+
+            _sendToClient(JsonRpcHelper.CreateSuccessResult(id, new SemanticTokensResponse() { Data = encodedTokens }));
+        }
+
+        private static ICollection<TokenType> ParseTokenTypesToSkipParam(string rawTokenTypesToSkipParam)
```

Comment:
>ICollection [](http://example.com/codeflow?start=23&length=11)

Again, why ICollection instead of IEnumerable? Just curious, might be something I don't know.

Reply:
I am using a Contains method in the places where this is used. I know there's where and other method but simple Contains does that trick. This is a hashset. I know there's also Contains on IEnumerable but not sure how it works and if it performs a linear search. 
<img width="271" alt="image" src="https://user-images.githubusercontent.com/25670945/234350823-ce81fcf0-bc8b-45f1-931a-58097856b86e.png">

Reply:
Resolving this. Let me know if you have any other concerns

Reply:
Yes, in this case if you're using this as a Set, I'd suggest returning ISet (or even just HashSet, since it's a private method)

Reply:
Sadly, IReadOnlySet isn't in .netstandard2.0 :(

Reply:
Guess i am used to returning everything as ICollection. But agreed i should be careful with my return types.